internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,497 @@
+package jobmanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/edulinq/autograder/internal/config"
+	"github.com/edulinq/autograder/internal/lockmanager"
+	"github.com/edulinq/autograder/internal/log"
+	"github.com/edulinq/autograder/internal/timestamp"
+	"github.com/edulinq/autograder/internal/util"
+)
+
+// Job output is locked on two levels:
+//   - The object level
+//   - The field level
+//
+// Locking on the JobID gives access to all fields.
+// When locking on a specific field, the JobID lock must be acquired first.
+const (
+	JOB_OUTPUT_LOCK_FORMAT     = "%s::%s"
+	JOB_OUTPUT_ERROR           = "error"
+	JOB_OUTPUT_WORK_ERRORS     = "work-errors"
+	JOB_OUTPUT_RESULT_ITEMS    = "result-items"
+	JOB_OUTPUT_REMAINING_ITEMS = "remaining-items"
+	JOB_OUTPUT_RUN_TIME        = "run-time"
+)
+
+// JobOptions contains user level options for a Job.
+// These options allow for optional context cancellation of the Job.
+type JobOptions struct {
+	// Don't save anything.
+	DryRun bool `json:"dry-run"`
+
+	// Remove any existing records before running the job.
+	OverwriteRecords bool `json:"overwrite-records"`
+
+	// Wait for the entire job to complete and return all results.
+	WaitForCompletion bool `json:"wait-for-completion"`
+
+	// Return a copy of an intermediate result.
+	// Will not be able to access final output.
+	// If WaitForCompletion is true, this option is overwritten.
+	ReturnCopy bool `json:"return-copy"`
```

Comment:
This doesn't feel like it should be a user-level option.
We have talked before about separating the structs with these user vs system level options.
Regardless of your decision here, this should probably at least not be exported in JSON.